src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1559 @@
+}
+
+/* This callback is triggered at regular intervals as long as a transaction is ongoing
+ * or a client has a TrustList open for reading or writing.
+ * If the session is no longer active, the current transaction is cancelled,
+ * and all open file handles are closed. */
+static void
+checkSessionActive(UA_Server *server, void *data) {
+    UA_GDSTransaction transaction = server->transaction;
+    UA_Boolean removeCallback = false;
+    if(transaction.state != UA_GDSTRANSACIONSTATE_FRESH) {
+        UA_Boolean foundSession = false;
+        session_list_entry *session;
+        LIST_FOREACH(session, &server->sessions, pointers) {
+            if(UA_NodeId_equal(&session->session.sessionId, &transaction.sessionId)) {
+                foundSession = true;
+                break;
+            }
+        }
+        if(!foundSession) {
+            UA_LOG_INFO(server->config.logging, UA_LOGCATEGORY_SERVER,
+                "Session with an open transaction has ended. "
+                "The transaction has been discarded.");
+            UA_GDSTransaction_clear(&server->transaction);
+            removeCallback = true;
+        }
+    } else {
+        removeCallback = true;
+    }
+
+    for(size_t i = 0; i < fileInfoContextSize; i++) {
+        UA_FileInfo fileInfo = fileInfoContext[i].fileInfo;
+
+        if(fileInfo.openCount == 0)
+            continue;
+
+        removeCallback = false;
+
+        UA_CertificateGroup *certGroup = getCertGroup(server, &fileInfoContext[i].certificateGroupId);
+
+        UA_FileContext *fileContext, *fileContextTmp;
+        LIST_FOREACH_SAFE(fileContext, &fileInfo.fileContext, listEntry, fileContextTmp) {
+            UA_Boolean foundSession = false;
+            session_list_entry *session;
+            LIST_FOREACH(session, &server->sessions, pointers) {
+                if(UA_NodeId_equal(&session->session.sessionId, &fileContext->sessionId)) {
+                    foundSession = true;
+                    break;
+                }
+            }
+
+            if(foundSession)
+                continue;
+
+            UA_LOG_INFO(server->config.logging, UA_LOGCATEGORY_SERVER,
+                    "Session with an open trust list has ended. "
+                        "All file handlers for the open trust lists have been closed.");
+
+            LIST_REMOVE(fileContext, listEntry);
+            fileInfoContext[i].fileInfo.openCount -= 1;
+
+            UA_ByteString_delete(fileContext->file);
+            UA_free(fileContext);
+
+            /* Updating OpenCount Varialbe in the information model */
+            UA_LOCK(&server->serviceMutex);
+            writeOpenCountVariable(server, certGroup);
+            UA_UNLOCK(&server->serviceMutex);
+        }
+    }
+
+    if(removeCallback) {
+        UA_Server_removeCallback(server, checkSessionCallbackId);
+        checkSessionCallbackId = 0;
+    }
+}
+
+static UA_Boolean
+compareThumbprint(const UA_String *str1, const UA_String *str2) {
```

Comment:
You can reuse `UA_String_equal_ignorecase`
https://github.com/open62541/open62541/blob/cd0aa1a987231b1b3cdf8145c30ce958f91b9fe6/src/util/ua_util_internal.h#L320